### ISSUE TITLE
Inconsistent validation metrics: Console output vs metrics.to_csv() results

### ISSUE THREAD

**Issue by DP1701** [ultralytics/ultralytics#21170]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello everyone,

I have written a small training script that performs a validation with the test dataset at the end and also saves the results in a CSV file. What I noticed is that the values in the CSV file do not match the values that are displayed in the console. It concerns the mAP values. Here is the script:

```Python
from ultralytics import YOLO
import os

# Load a model
model = YOLO("exported_last.pt")  # load a pretrained model (recommended for training)

# Paramter seettings
data="/raid/USERDATA/pawlodwp/dataset/labeled/Construction_Machines_v1/data.yaml"
batch=192 # 288 yolo11m
epochs=150 
imgsz=640 
device=[2,3]
cache=True
patience=20
project="yolo11l_machine_worker_pretrained"

# Train the model
results = model.train(data=data,
                      batch=batch, 
                      epochs=epochs, 
                      imgsz=imgsz, 
                      device=device, 
                      name="train",
                      project=project,
                      patience=patience,
                      cache=cache)
print("............")
metrics = model.val(data=data,
                    batch=batch,
                    imgsz=imgsz,
                    device=device,
                    name="test",
                    project=project, 
                    split="test")
print("............")

csv_str = metrics.to_csv()
csv_path = os.path.join(project, "validation_metrics.csv")

with open(csv_path, "w") as f:
    f.write(csv_str)
```

This is the output in the console:

![Image](https://github.com/user-attachments/assets/463f99e9-5f65-4020-87e8-5bad454f1f99)

This is the CSV:

[validation_metrics.csv](https://github.com/user-attachments/files/20878082/validation_metrics.csv)

I would like to have the same values in the same output in the CSV file. Including the average values across all classes (all) as in the console output. Does anyone have an idea?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DP1701, thank you for bringing this up and for your detailed report! 🚀 We appreciate your interest in Ultralytics and your clear description of the behavior you’re seeing.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us debug the issue. Sharing your full script, sample outputs, and any custom configurations will help our team quickly investigate the inconsistency you observed between the console mAP values and those saved in your CSV.

An Ultralytics engineer will review your report soon and provide more specific guidance.

In the meantime, we recommend checking our [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as answers to many common questions.

Join the Ultralytics community for real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, in-depth discussions on [Discourse](https://community.ultralytics.com/), or knowledge sharing on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

Please upgrade to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue is not already resolved in the latest release:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will assist you here soon!

**Comment by RizwanMunawar** [comment]
@DP1701 Hi, Thanks for sharing the details. For CSV, HTML, and other formats, we consider mAP values for mAP@50, mAP@75, and in terminal, these values are displayed for each class, which is (average precision, not mean average precision).

**Comment by DP1701** [comment]
Thanks for the explanation. I was a little confused and didn't read the mAP values in the console as AP values.

**Comment by DP1701** [comment]
@RizwanMunawar Hi, how can I get the information how many images and instances per class have been evaluated? Like in the picture below:

![Image](https://github.com/user-attachments/assets/4db1127a-6a71-4971-96cd-4cb369a5fb52)

**Comment by RizwanMunawar** [comment]
@DP1701 Well, this is something we dont support for now. I will see if this is possible directly with the validation results export.

**Comment by RizwanMunawar** [comment]
@DP1701 Hi, I just opened the [PR](https://github.com/ultralytics/ultralytics/pull/21203), with this you will be able to see the Images/Instances and additionally per class average precision. Thanks

**Comment by DP1701** [comment]
Wow, thank you! 
Then I don't have to think of a workaround to get the data.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #21203, which aligns the metrics summary method to match the validation logs. This update brings several improvements:

- Validation result summaries are now clearer and more consistent, making it easier to interpret and export YOLO model metrics.
- Metric keys have been renamed for clarity, and more detailed per-class information (like the number of images and instances) is now available.
- Classification metrics have standardized keys (`top1_acc`, `top5_acc`), and some less-used metrics have been removed from the summary.
- Documentation now includes a direct Colab link to help users export validation results to different formats such as DataFrame, CSV, and SQL.

Thanks to the efforts of the PR author and contributors, these changes should streamline metric analysis and export workflows.

To test if this resolves your issue, try installing the latest code:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you can wait for the next official PyPI release, which will include these updates.

Please let us know if this update resolves your issue or if you notice any further concerns. Thank you for reporting this, and any additional feedback is always welcome! 🙏
